### PR TITLE
Issue 6223: SLTS - Enable loading third party storages.

### LINKED ISSUE #6230: SLTS - Cannot load 3rd party storages
**Describe the bug**
The storageImplementation field in ServiceConfig is a strongly typed enum.
This is unintentional and prevents loading of third party storages.
The field should be a string field/property

**To Reproduce**
Try specifying and loading any storage outside of Pravega tree

**Screenshots**
None

**Additional information**
None

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
Signed-off-by: Sachin Joshi <sachin.joshi@emc.com>

**Change log description**  
SLTS - Enable loading third party storages.

**Purpose of the change**  
Fixes #6230 #4993

**What the code does**  
- The storageImplementation field in ServiceConfig is a strongly typed enum.
- This is unintentional and prevents loading of third party storages.
- The field should be a string field/property

**How to verify it**  
All tests should pass

### CODE REVIEW

**Review by eolivelli — CHANGES_REQUESTED** [review]
The patch makes sense to me.

We will have to add some minimal test in order to see it bootstrapping some non system LTS implementation.

I will try to run my custom implementation

**Comment by eolivelli** [comment]
One question:
where do I put the jars for the LTS implementation ?

**Comment by sachin-j-joshi** [comment]
The storage implementations are loaded using `ServiceLoader` (https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)

As long as implementation jar is on the class path it should work.

In fact, there is nothing special about existing bindings other than they are part of the same repo and when we create docker image it puts bindings jar in same directory. But otherwise they use the same mechanism to load.

(The mechanism to load storage factory was already implemented in 0.8 timeframe however this specific bug in config parsing prevented passing arbitrary value to loader.)

Here are the steps to follow. 
**Step 1:**  Add resource file to indicate the class that implements Storage Factory creator
Example - https://github.com/pravega/pravega/blob/master/bindings/src/main/resources/META-INF/services/io.pravega.segmentstore.storage.StorageFactoryCreator

**Step 2:** Implement the StorageFactoryCreator that creates the factory for storage instances
Example  -https://github.com/pravega/pravega/blob/master/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorageFactoryCreator.java

**Step 3:** Implement factory 
https://github.com/pravega/pravega/blob/master/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemSimpleStorageFactory.java
Add custom configuration.
https://github.com/pravega/pravega/blob/master/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorageConfig.java


(https://github.com/pravega/pravega/blob/master/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorage.java is where ChunkStorage is implemented.)

**Review by eolivelli — CHANGES_REQUESTED** [review]
Lgtm

So there is no need for additional testing.

**Review by eolivelli — APPROVED** [review]

**Review by andreipaduroiu — COMMENTED** [review]
This looks OK to me, but how is a user to know how to load an external implementation.

At least one example, even if in a comment, on how to do that and what to pass into this config value would be great. I suggest we write this in the `StorageConfig` class (where we defined this property) and in `config.properties` where we list this property.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]
.

**Comment by eolivelli** [comment]
> At least one example, even if in a comment, on how to do that and what to pass into this config value would be great. I suggest we write this in the `StorageConfig` class (where we defined this property) and in `config.properties` where we list this property.

I second this. good idea @andreipaduroiu

**Comment by sachin-j-joshi** [comment]
Updated comments - https://github.com/pravega/pravega/pull/6231/commits/9f0d53b6a430d24cdc1d3fa943eea9a1bbdd971b

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorage.java [hunk 1]
```diff
@@ -49,8 +49,22 @@
...
+ *     <li> Implement {@link ChunkStorage}. It is recommended that the implementations should extend {@link AsyncBaseChunkStorage}
+ *     or {@link BaseChunkStorage}.</li>
+ *     <li> Implement {@link io.pravega.segmentstore.storage.SimpleStorageFactory}.  </li>
+ *     <li> Implement {@link io.pravega.segmentstore.storage.StorageFactoryCreator}.
+ *     <ul>
+ *          <li>Return {@link io.pravega.segmentstore.storage.StorageFactoryInfo} object containing name to use to identify.
+ *          This identifier is used in config.properties file to identify implementation to use.</li>
+ *          <li> The bindings are loaded using ServiceLoader (https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)</li>
+ *          <li> Add resource file named "io.pravega.segmentstore.storage.StorageFactoryCreator" to the implementation jar.
+ *          File contains name of the class implementing {@link io.pravega.segmentstore.storage.StorageFactoryCreator}.</li>
```
Comment (eolivelli) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorage.java:64]: typo: "File contains name" -> "That file contains the name"

**Comment by eolivelli** [comment]
@andreipaduroiu can you please take another look ?

cc @diegosalvi you are interested in this patch as well

**Review by andreipaduroiu — APPROVED** [review]

**Pull request merged**